### ISSUE TITLE
After Entry Save Hook lies

### ISSUE THREAD

**Issue by blairliikala** [ExpressionEngine/ExpressionEngine#3852]
When using the `after_channel_entry_save` hook, I would expect the data to have been already written to the database.  After testing, **custom fields** are not written to the database until **after** this hook runs.  Native fields **are** written, such as title and date.  This affects add-on functionality when trying to retrieve or send data outside of the add-on, such as calling a template.

Steps to Test/Reproduce
- Create an addon that listens for the hook.
- Add `dd()` or `exit` in the add-on callback function to stop code.
- Try editing the title through the usual CP form, then look in the database for changes.  It should change.
- Edit a custom field, and note no changes in the database.  It should not change.

Solution:
By the time `after_channel_entry_save` is called, all entry data should already be written to the database.

EE 7.3.13, php 8.1.

**Comment by shbchk** [comment]
Feel like this is the perfect place for this feature request:

It would be great for Weebhooks users to have an option to compare values from previous state with the new state and take actions based on results of that comparison.

**Comment by blairliikala** [comment]
@shbchk Yes, and I have it somewhat ready in the background.  However to see if data changed in a hook is a bit tricky.  Ideally **every** hook would include the before and after data, and a **complete** payload of the object.  Then it is easy to see what changed.

In reality only a few hooks support the `modified` argument.  Those that do only show a change for basic fields, so for the entry model, Grid, Relationships, and Fluid are not included.  Said another way, if you change a grid field, then it does not reflect in the `modified` argument.  In general throughout EE, the entry model lacks here, and I think there are plans to fix that.  There's probably some crafty work-arounds, but I want to be careful about introducting brittle code.  If there's a better way I'm all ears!

**Comment by intoeetive** [comment]
@blairliikala thank you for reporting this.

On your tests, did you experience the issue with complex fields, such as Grid or Relationship, or even the simple fields such as Text?

**Comment by blairliikala** [comment]
Complex feeds show up in the callback objects as `null` of course, so that creates its own complications here.

- Fluid updates in the database correctly with the exception when deleting the last fluid field (1 field to 0 fields).
- Grid updates in the database correctly.
- Relationships update in the database correctly.
- Textarea and text fields...etc do not update in the db with this hook.

**Comment by gb-lin** [comment]
I encountered the same thing with my custom fieldtype and I found that the issue is $object->emit('afterUpdate', $backup); in Update.php. I think it kind of make sense for afterUpdate to fire after afterSave and by moving the saving sequence from afterSave to afterUpdate my custom values are saved correctly. But I didn't read enough to understand why the $backup is needed.
